pkg/diff/sql_generator.go

```diff
@@ -842,32 +875,90 @@ func (t *tableSQLVertexGenerator) GetAddAlterDependencies(table, _ schema.Table)
 			mustRun(t.GetSQLVertexId(table), diffTypeAddAlter).after(buildTableVertexId(table.ParentTableName), diffTypeAddAlter),
 		)
 	}
-	return deps
+	return deps, nil
+}
+
+func getDangerousNotNullAlters(alteredCols []columnDiff, newSchemaCCs []schema.CheckConstraint, oldSchemaCCs []schema.CheckConstraint) []columnDiff {
+	var ccs []schema.CheckConstraint
+	ccs = append(ccs, newSchemaCCs...)
+	ccs = append(ccs, oldSchemaCCs...)
+
+	// A dangerous not null alter is an alter that changes a column from nullable to not null, but does not have a
+	// valid NOT NULL check constraint backing it
+	safeColsByName := make(map[string]bool)
+	for _, cc := range ccs {
+		if isValidNotNullCC(cc) {
+			safeColsByName[cc.KeyColumns[0]] = true
+		}
+	}
+
+	var dangerousNotNullAlters []columnDiff
+	for _, colDiff := range alteredCols {
+		if colDiff.old.IsNullable && !colDiff.new.IsNullable && !safeColsByName[colDiff.new.Name] {
+			dangerousNotNullAlters = append(dangerousNotNullAlters, colDiff)
+		}
+	}
+
+	return dangerousNotNullAlters
+}
+
+var (
+	// isValidNotNullCCRegex is a regex that matches a valid NOT NULL check constraint. It's covers the simplest case,
+	// so we might have to improve it in the future.
+	// Notably, a false positive is much worse than a false negative because a false negative will result in an unnecessary
+	// check constraint being built, but a false positive will result in a column being changed to `NOT NULL` without
+	// a backing check constraint or a warning to the user. We should be very conservative with this regex.
+	isNotNullCCRegex = regexp.MustCompile(`^\(*((".*")|(\S*)) IS NOT NULL\)*$`)
+)
+
+func isValidNotNullCC(cc schema.CheckConstraint) bool {
+	if len(cc.KeyColumns) != 1 {
+		return false
+	}
+	if !cc.IsValid {
+		return false
+	}
+	return isNotNullCCRegex.MatchString(cc.Expression)
+}
+
+func buildTempNotNullConstraint(colDiff columnDiff) (schema.CheckConstraint, error) {
+	uuid, err := uuid.NewRandom()
+	if err != nil {
+		return schema.CheckConstraint{}, fmt.Errorf("generating uuid: %w", err)
+	}
+	return schema.CheckConstraint{
+		Name:               fmt.Sprintf("not_null_%s", uuid.String()),
```

Comment:
Great point!

Reply:
I want to update the temporary indexes to use this pattern, but I'm going to save that for a subsequent PR, since I also am going to switch to the UUID's to base64 encoded to make them smaller.